src/grift/insert-casts.rkt

```diff
@@ -252,7 +289,7 @@
             [else
              (define cst (mk-cast e-src lbl e DYN-TYPE (MRef DYN-TYPE)))
              (MBoxCastedRef cst DYN-TYPE)]))]
-       [(MRef t)
+       [(or (MRef t) (and (Mu _) (app unfold-mu (MRef t))))
```

Comment:
If I am not, it is a bug. The cleanest fix might be to call what is currently called `resolve` on the type and then just do the match for the concrete constructor. It will simplify all of this code.